src/Features/Core/Portable/CodeRefactorings/SyncNamespace/AbstractChangeNamespaceService.cs

```diff
@@ -0,0 +1,592 @@
+        /// in the new node might be qualified with this new namespace instead.
+        /// </summary>
+        /// <param name="reference">A reference to a type declared inside the namespce to be changed, which is calculated 
+        /// based on results from `SymbolFinder.FindReferencesAsync`.</param>
+        /// <param name="newNamespaceParts">If specified, the namespace of original reference will be replaced with given 
+        /// namespace in the replacement node.</param>
+        /// <param name="old">The node to be replaced. This might be an ancestor of original </param>
+        /// <param name="new">The replacement node.</param>
+        public abstract bool TryGetReplacementReferenceSyntax(SyntaxNode reference, ImmutableArray<string> newNamespaceParts, ISyntaxFactsService syntaxFacts, out SyntaxNode old, out SyntaxNode @new);
+    }
+
+    internal abstract class AbstractChangeNamespaceService<TNamespaceDeclarationSyntax, TCompilationUnitSyntax, TMemberDeclarationSyntax>
+        : AbstractChangeNamespaceService
+        where TNamespaceDeclarationSyntax : SyntaxNode
+        where TCompilationUnitSyntax : SyntaxNode
+        where TMemberDeclarationSyntax : SyntaxNode
+    {
+        private static readonly char[] s_dotSeparator = new[] { '.' };
+
+        protected abstract TCompilationUnitSyntax ChangeNamespaceDeclaration(
+            TCompilationUnitSyntax root, ImmutableArray<string> declaredNamespaceParts, ImmutableArray<string> targetNamespaceParts);
+
+        protected abstract SyntaxList<TMemberDeclarationSyntax> GetMemberDeclarationsInContainer(SyntaxNode compilationUnitOrNamespaceDecl);
+
+        protected static SyntaxAnnotation WarningAnnotation { get; }
+            = CodeActions.WarningAnnotation.Create(
+                FeaturesResources.Warning_colon_changing_namespace_may_produce_invalid_code_and_change_code_meaning);
+
+        /// <summary>
+        /// This code action tries to change the name of the namespace declaration to 
+        /// match the folder hierarchy of the document. The new namespace is constructed 
+        /// by concatenating the default namespace of the project and all the folders in 
+        /// the file path up to the project root.
+        /// 
+        /// For example, if he default namespace is `A.B.C`, file path is 
+        /// "[project root dir]\D\E\F\Class1.cs" and declared namespace in the file is
+        /// `Foo.Bar.Baz`, then this action will change the namespace declaration
+        /// to `A.B.C.D.E.F`. 
+        /// 
+        /// Note that it also handles the case where the target namespace or declared namespace 
+        /// is global namespace, i.e. default namespace is "" and the file is located at project 
+        /// root directory, and no namespace declaration in the document, respectively.
+        /// </summary>
+        public override async Task<Solution> ChangeNamespaceAsync(
```

Comment:
Yes. In addition, there's still some dependency between service and refactoring right now, which needs to be fixed (e.g. service assumes the shape of syntax tree for documents passed in). I delayed that change to get this part merged first so it will be easier to review later.